pkg/segment/segment.go

```diff
@@ -0,0 +1,153 @@
+	"io/ioutil"
+	"net"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+// Writekey will be the API key used to send data to the correct source on Segment
+var WriteKey = "4xGV1HV7K2FtUWaoAozSBD7SNCBCJ65U"
+
+type Client struct {
+	segmentClient     analytics.Client
+	config            *preference.PreferenceInfo
+	telemetryFilePath string
+}
+
+func NewClient(config *preference.PreferenceInfo) (*Client, error) {
+	homeDir, _ := os.UserHomeDir()
+	return newCustomClient(config,
+		filepath.Join(homeDir, ".redhat", "anonymousId"),
+		analytics.DefaultEndpoint,
+	)
+}
+
+func newCustomClient(config *preference.PreferenceInfo, telemetryFilePath string, segmentEndpoint string) (*Client, error) {
+	client, err := analytics.NewWithConfig(WriteKey, analytics.Config{
+		Endpoint: segmentEndpoint,
+		DefaultContext: &analytics.Context{
+			IP: net.IPv4(0, 0, 0, 0),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		segmentClient:     client,
+		config:            config,
+		telemetryFilePath: telemetryFilePath,
+	}, nil
+}
+
+// Close client connection and send the data
+func (c *Client) Close() error {
+	return c.segmentClient.Close()
+}
+
+// Upload prepares the data to be sent to segment and send it once the client connection closes
+func (c *Client) Upload(action string, duration time.Duration, err error) error {
+	// if the user has not consented for telemetry, return
+	if !c.config.GetConsentTelemetry() {
+		return nil
+	}
+
+	// obtain the anonymous ID
+	anonymousID, uerr := getUserIdentity(c.telemetryFilePath)
+	if uerr != nil {
+		return uerr
+	}
+
+	// queue the data that helps identify the user on segment
+	if err := c.segmentClient.Enqueue(analytics.Identify{
+		AnonymousId: anonymousID,
+		Traits:      addConfigTraits(c.config, traits()),
+	}); err != nil {
+		return err
+	}
+
+	// add information to the data
+	properties := analytics.NewProperties()
+	// TODO: add other properties when required
+	properties = properties.Set("version", "odo "+version.VERSION+" ("+version.GITCOMMIT+")").
+		Set("success", err == nil).
+		Set("duration(ms)", duration.Milliseconds()).
+		Set("tty", RunningInTerminal())
+	if err != nil {
+		properties = properties.Set("error", SetError(err)).Set("error-type", errorType(err))
+	}
+
+	// queue the data that has telemetry information
+	return c.segmentClient.Enqueue(analytics.Track{
+		AnonymousId: anonymousID,
+		Event:       action,
+		Properties:  properties,
+	})
+}
+
+// addConfigTraits add more information to be sent to segment
+// Note: This currently acts as a placeholder
+func addConfigTraits(c *preference.PreferenceInfo, in analytics.Traits) analytics.Traits {
+	// TODO: add more traits later
+	return in
+}
+
+// getUserIdentity returns the anonymous ID if it exists, else creates a new one
+func getUserIdentity(telemetryFilePath string) (string, error) {
+	var id []byte
+
+	if err := os.MkdirAll(filepath.Dir(telemetryFilePath), 0750); err != nil {
```

Comment:
please check how other functions that use MkdirAll etc.. use the afero for testing 🙂 .

That would make writing unit test for this code easy